src/clang-cpp-frontend/clang_cpp_convert.cpp

```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
 
   return false;
 }
+
+bool clang_cpp_convertert::annotate_class_field_access(
+  const clang::FieldDecl *field,
+  struct_typet::componentt &comp)
+{
+  std::string access;
+  if(get_access_from_decl(field, access))
+    return true;
+
+  // annotate access in component
+  comp.set_access(access);
+  return false;
+}
+
+bool clang_cpp_convertert::get_access_from_decl(
+  const clang::Decl *decl,
+  std::string &access)
+{
+  switch(decl->getAccess())
+  {
+  case clang::AS_public:
+  {
+    access = "public";
+    break;
+  }
+  case clang::AS_private:
+  {
+    access = "private";
+    break;
+  }
+  case clang::AS_protected:
+  {
+    access = "protected";
+    break;
+  }
+  default:
+  {
+    log_error("Unknown specifier returned from clang in {}", __func__);
+    return true;
+  }
+  }
+
+  return false;
+}
+
+bool clang_cpp_convertert::annotate_cpp_methods(
+  const clang::CXXMethodDecl *cxxmdd,
+  exprt &new_expr)
+{
+  code_typet &component_type = to_code_type(new_expr.type());
+
+  // annotate ctor and dtor return type
+  if(cxxmdd->getKind() == clang::Decl::CXXDestructor)
+  {
+    typet rtn_type("destructor");
+    component_type.return_type() = rtn_type;
+  }
+  if(cxxmdd->getKind() == clang::Decl::CXXConstructor)
+  {
+    typet rtn_type("constructor");
+    component_type.return_type() = rtn_type;
+  }
+
+  // annotate parent
+  std::string parent_class_name = getFullyQualifiedName(
+    ASTContext->getTagDeclType(cxxmdd->getParent()), *ASTContext);
+  std::string parent_class_id = tag_prefix + parent_class_name;
+  component_type.set("#member_name", parent_class_id);
```

Comment:
do we need this information?

Also, instead of exporting getFullyQualifiedName, you should call `get_decl(cxxmdd->getParent())` and access its name.

I fail to see why we need this code tho.

Reply:
>Also, instead of exporting getFullyQualifiedName, you should call get_decl(cxxmdd->getParent()) and access its name.

well I thought of using `get_decl(cxxmdd->getParent(), new_expr)` ...  but it would appear to go through lots of recursions, seems to be an overkill and might be quite head scratching when we need to debug in the future (e.g. dealing with 25 frames in gdb). Obviously, here we just want the to get the id and annotate `member_name`. Using the getFullyQualifiedName function appears to be a neat way to do so.